supermarq-benchmarks/supermarq/qcvv/base_experiment.py

```diff
@@ -0,0 +1,462 @@
+
+            experiment = ExampleExperiment(<<args/kwargs>>)
+
+    #. Run the experiment on the desired target. This can either be a custom simulator
+       or a real device name. For example
+
+        .. code::
+
+            noise_model = cirq.depolarize(p=0.01, n_qubits=1)
+            target = cirq.DensityMatrixSimulator(noise=noise_model)
+
+            experiment.run(target=target, <<args/kwargs>>)
+
+    #. Then we analyse the results. If the target was a local simulator this will be available as
+       soon as the :code:`run()` method has finished executing. On the other hand if a real device
+       was accessed via Superstaq then it may take time for the data to be available from the
+       server. The :code:`collect_data()` will return :code:`True` when all data has been collected
+       and is ready to be analysed.
+
+       .. code::
+            if self.collect_data():
+                results = experiment.analyse_results(<<args>>)
+
+    #. The final results of the experiment will be stored in the :code:`results` attribute as a
+       :class:`~typing.NamedTuple` of values, while all the data from the experiment will be
+       stored in the :code:`raw_data` attribute as a :class:`~pandas.DataFrame`. Some experiments
+       may include additional data attributes for data generated during the analysis.
+
+        .. code::
+
+            results = experiment.results
+            data = experiment.raw_data
+
+    .. warning::
+        Note that each time the :code:`run()` method is called the
+        previous jobs, results and data are overwritten.
+
+    When implementing a new experiment, 4 methods need to be implemented:
+
+    #. :meth:`build_circuits`: Given a number of circuits and an iterable of the different numbers
+        of layers to use, return a list of :class:`Sample` objects that need to be sampled during
+        the experiment.
+
+    #. :meth:`process_probabilities`: Take the probability distribution over the
+        computational basis resulting from running each circuit and combine the relevant details
+        into a :class:`pandas.DataFrame`.
+
+    #. :meth:`analyse_results`: Analyse the data in the :attr:`raw_data` dataframe and return a
+        :class:`~typing.NamedTuple`-like object containing the results of the experiment.
+
+    #. :meth:`plot_results`: Produce any relevant plots that are useful for understanding the
+        results of the experiment.
+
+    """
+
+    def __init__(
+        self,
+        num_qubits: int,
+        **kwargs: Any,
+    ) -> None:
+        """Args:
+        num_qubits: The number of qubits used during the experiment. Most subclasses
+            will determine this from their other inputs.
+        """
+        self.qubits = cirq.LineQubit.range(num_qubits)
+        """The qubits used in the experiment."""
+
+        self._raw_data: pd.DataFrame | None = None
+        "The data generated during the experiment"
+
+        self._results: NamedTuple | None = None
```

Comment:
I did have a go at this originally, but the results from each different type of experiment have completely different names so their didn't seem to be any benefit to subclassing an empty dataclass. I could have use a dict but wanted immutability and strong typing so went with a NamedTuple.

Reply:
that makes sense, though imo it would be worth it even if the parent result class just contained a couple boilerplate items like "target"/experiment type/etc. We could used `dataclasses.dataclass(frozen=True)` to enforce immutability

imo it would be nice to have a single abstract result type we could type check against. this might be silly though so your call - im fine with it as is if you prefer

Reply:
A good idea to add some boiler plate items, which then makes dataclasses the best option. Thanks - I've updated